File: 907-koko-eating-bananas/koko-eating-bananas.js
/**
 * @param {number[]} piles
 * @param {number} h
 * @return {number}
 */
var minEatingSpeed = function (piles, h) {
    let low = 0, high = Math.max(...piles), ans = Infinity

    while (low <= high) {
        let mid = Math.floor((low + high) / 2)
        let totalHours = calculateTotalHours(piles, mid)
        if (totalHours <= h){
             high = mid - 1
             ans = mid
        }else {
            low = mid + 1
        }
    }

    return ans;


};

const calculateTotalHours = (arr, hourly) => {
    let totalHours = 0
    for (let num of arr) {
        totalHours += Math.ceil(num / hourly)
    }
    return totalHours;

}

// claude chat link
// https://claude.ai/chat/9dfc5571-206f-49f8-914b-30f030d026f6

File: 907-koko-eating-bananas/koko-eating-bananas.py
import math

class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        low = 1
        high = max(piles)
        ans = float('inf')

        while low <= high:
            mid = (low + high) // 2
            # total_hours = sum(math.ceil(pile / mid) for pile in piles)
            total_hours = self.calculate_hours(piles, mid)

            if total_hours <= h:
                high = mid-1
                ans = mid
            else:
                low = mid + 1
                

        return ans
    
    def calculate_hours(self, arr, hourly):
        total_hours = 0
        for num in arr:
            total_hours += math.ceil(num/hourly)
        return total_hours
